### ISSUE TITLE
Build failure for 1.0.3627

### ISSUE THREAD

**Issue by mjhammel** [intel/intel-graphics-compiler#131]
Trying to build igc with the following command.

> 
> cd $$@/build && \
>         cmake \
>               -DCMAKE_BUILD_TYPE=Release \
>               -DCMAKE_SYSTEM_PROCESSOR=x86_64 \
>               -DCMAKE_TOOLCHAIN_FILE=$(LOCAL_TOOLS)/bin/$($1_ARCH).cmake \
>               -DCMAKE_CROSSCOMPILING=TRUE \
>               -DIGC_OPTION__INCLUDE_IGC_COMPILER_TOOLS=OFF \
>               -DIGC_BUILD__SHOW_CFLAGS=ON \
>               -DCMAKE_CXX_FLAGS="-Wno-dev" \
>               -DCMAKE_C_FLAGS="-Wno-dev" \
>               -DCCLANG_BUILD_INTREE_LLVM=TRUE \
>               -DHAVE_POSIX_REGEX=0 \
>               -DHAVE_STEADY_CLOCK=0 \
>               LD="$($1_ARCH)-gcc -shared" \
>               CC="$($1_ARCH)-gcc" \
>               ../igc-$(IGC_VERSION)/IGC
>     $(MAKE) -C $$@/build VERBOSE=1

I run into this error:

> make -f llvm/build/src/include/llvm/IR/CMakeFiles/intrinsics_gen.dir/build.make llvm/build/src/include/llvm/IR/CMakeFile
> s/intrinsics_gen.dir/build
> make[3]: Entering directory '/home/qsysbuilder/src/qsc/qsys2/src/os/x86_64-linux-gnu/igc/build'
> make[3]: *** No rule to make target 'NATIVE/bin/llvm-tblgen', needed by 'llvm/build/src/include/llvm/IR/Attributes.inc'.
>   Stop.
> make[3]: Leaving directory '/home/qsysbuilder/src/qsc/qsys2/src/os/x86_64-linux-gnu/igc/build'
> CMakeFiles/Makefile2:1472: recipe for target 'llvm/build/src/include/llvm/IR/CMakeFiles/intrinsics_gen.dir/all' failed
> make[2]: *** [llvm/build/src/include/llvm/IR/CMakeFiles/intrinsics_gen.dir/all] Error 2
> 

If I need to bump to newer version (or back to older one) that's fine.
Let me know if you need any additional information.  Thanks.

**Comment by mjhammel** [comment]
I think I got past this one, and another like it, by adding the following to the cmake command:

> 
> -DLLVM_TABLEGEN=$(QSYS_TOP)/os/$$@/build/NATIVE/llvm/build/src/bin/llvm-tblgen \
> -DCLANG_TABLEGEN=$(QSYS_TOP)/os/$$@/build/NATIVE/llvm/build/src/bin/clang-tblgen \

That's a fully qualified path to the tablgen utilities that are being built by the IGC bulid.  For some reason the full build can't reference them without this, but I know where they'll be built to so can set them before the build starts.  At least I think this works - I haven't tried a full build yet, but adding these got me past the build failures without restarting the whole build.

**Comment by mjhammel** [comment]
Next problem:  I seem to be missing "clang-tool" from the build.  I don't see it as a build artifact but the build relies on it.

[igc.txt](https://github.com/intel/intel-graphics-compiler/files/4458147/igc.txt)

**Comment by maleadt** [comment]
Running into the same issue, and it seems cross-compilation related. Setting CMAKE_CROSSCOMPILING to false (and running on a build platform that matches the target) results in a single build of these tools, not in the NATIVE folder, that are properly picked up.

Re. the clang-tool issue, I substituted the path to clang-tool like this: https://github.com/JuliaPackaging/Yggdrasil/blob/7e16fe57d100899f2b9045c0a54c2093bba5e757/L/libigc/build_tarballs.jl#L85-L92. However, it subsequently couldn't execute ocloc, see https://github.com/intel/opencl-clang/issues/91, so I disabled embedding the OpenCL headers like this: https://github.com/JuliaPackaging/Yggdrasil/blob/7e16fe57d100899f2b9045c0a54c2093bba5e757/L/libigc/build_tarballs.jl#L27-L29 (I only target oneAPI). Sadly, that has lead me to https://github.com/intel/intel-graphics-compiler/issues/128...

Bottom line, it looks like igc isn't cross-compilable right now.

**Comment by mjhammel** [comment]
Was your testing with the latest tagged version or an earlier version?  I started with 1.0.3627, tried 1.0.3698, then went as far back as the released version (1.0.2714) but had different build problems with each.  Was wondering if there might be an earlier release I could try that might build.  FWIW, I'm building igc for use wiht dldt's GPU plugin.

**Comment by maleadt** [comment]
I was using the 1.0.3586 release because that's the version recommended by the NEO release I was looking at.

**Comment by mjhammel** [comment]
I've managed to get a reproducable build up to the missing _clang-tool_ issue using the latest tagged version (1.0.3698).  To do this I have to manually build _llvm-tblgen_, _clang-tblgen_ and _linux_resource_linker_.  These are all GNU Make targets in the Makefile generated by cmake.  I also have to symlink build/NATIVE to the build tree itself.  This works for my case because the host and target systems are x86_64, but I don't think this would work for different host/target architectures.

So now my build looks like the following.

    mkdir -p $$@/build
    cd $$@/build && rm -f NATIVE && ln -s . NATIVE
    cd $$@/build && \
        cmake \
              -DCMAKE_BUILD_TYPE=Release \
              -DCMAKE_SYSTEM_PROCESSOR=x86_64 \
              -DCMAKE_TOOLCHAIN_FILE=$(LOCAL_TOOLS)/bin/$($1_ARCH).cmake \
              -DCMAKE_CROSSCOMPILING=TRUE \
              -DIGC_BUILD__SHOW_CFLAGS=ON \
              -DIGC_OPTION__ARCHITECTURE_TARGET=Linux64 \
              -DIGC_OPTION__ARCHITECTURE_HOST=Linux64 \
              -DLLVM_INCLUDE_UTILS=ON \
              -DLLVM_INSTALL_UTILS=ON \
              -DLLVM_BUILD_UTILS=ON \
              -DLLVM_TABLEGEN=$(QSYS_TOP)/os/$$@/build/NATIVE/llvm/build/src/bin/llvm-tblgen \
              -DCLANG_TABLEGEN=$(QSYS_TOP)/os/$$@/build/NATIVE/llvm/build/src/bin/clang-tblgen \
              -DCCLANG_BUILD_INTREE_LLVM=TRUE \
              -DHAVE_POSIX_REGEX=0 \
              -DHAVE_STEADY_CLOCK=0 \
              ../igc-$(IGC_VERSION)/IGC
    $(MAKE) -j$(SIMUL) -C $$@/build VERBOSE=1 llvm-tblgen
    $(MAKE) -j$(SIMUL) -C $$@/build VERBOSE=1 clang-tblgen
    $(MAKE) -j$(SIMUL) -C $$@/build VERBOSE=1 linux_resource_linker
    $(MAKE) -j$(SIMUL) -C $$@/build VERBOSE=1

Unfortunately there is no _clang-tool_ target in the generated Makefile.  My guess is that the tool target name is something different, but I don't know what clang-tool does so I can't (yet) determine what other target might suffice in its place.

**Comment by mjhammel** [comment]
Latest release (3698) didn't work either.  Has the same missing clang-tool error.  We decided to pass on the source build for now and will now try using the ubuntu binaries for dldt (which is what we were building that needed igc) since our target is an x86-64 and hope they just work.

**Comment by paigeale** [comment]
Issue is no longer present in latest compute runtime/ intel-graphics-compiler. Please reopen if issue persists
